winch/codegen/src/codegen/call.rs

```diff
@@ -11,15 +13,24 @@ pub(crate) struct FnCall<'a> {
     /// The total stack space needed for the callee arguments on the
     /// stack, including any adjustments to the function's frame and
     /// aligned to to the required ABI alignment.
-    pub total_arg_stack_space: u32,
+    total_arg_stack_space: u32,
     /// The stack space consumed by spilling registers placed as
     /// function arguments in the value stack.  This gets implicitly
     /// allocated by saving any live registers in the value stack, but
     /// we track it to make sure to reclaim that space once the
     /// function call ends.
-    pub spill_stack_space: u32,
+    spill_stack_space: u32,
+    /// The amount of memory in bytes consumed by the call. That is,
+    /// any memory entries that will be used as params and that exist
+    /// before any spills generated by saving live registers are
+    /// counted as memory entries that will be consumed by the call,
+    /// and thus we track this space to claim it once the function
+    /// call has ended. This field is analogous to
+    /// `spill_stack_space`, but it's kept separate for clarity and
+    /// debugging purposes.
```

Comment:
I definitely appreciate the docs here, but a few things are slightly unclear to me at least:

- The terminology of "memory entries" is confusing -- does this just mean "values on the stack"?
- The "That is, any memory entries ... " sentence is really hard to parse. I *think* what this is saying is: values already on the stack that are args?
- "Claim" means "reclaim" here?

Is it the case that `memory_consumed + spill_stack_space == total_arg_stack_space`? If so, could we use that alone to pop args off the stack? Fundamentally it seems like we shouldn't have to know where the args originally came from when we're cleaning up.

Reply:
> Is it the case that memory_consumed + spill_stack_space == total_arg_stack_space?

I see where you're coming from and I agree that it's a bit confusing, I think I can make it so that it's more clear and so that we have to deal with less terms.

But to give a bit more context:

* total_arg_stack_space is the total space that needs to be explicitly allocated at the callsite, and covers: the stack space needed for arguments plus any alignment needed for the caller's frame. 

* memory_consumed, represents the memory used by any `Memory` entries existing in the value stack at the moment of assigning those entries to the function arguments (in the case of a register argument, that memory will be loaded to the register used as argument, and in the case of a stack argument, that memory will be loaded to the scratch register and then stored to the particular argument slot). We could emit a `pop` instruction when doing this assignment and will potentially simplify things here in terms of what we are tracking, but I think it's more efficient to track the amount of memory used and reclaim it in the end.

* spill_stack_space is the space created to save any live registers.

Does that clarify things to you? 

As I write this, I believe we can simplify this by tracking all the memory used in a single field. Initially I thought that separating where things come from would make the code more understandable and easier to reason about, but maybe the opposite is better.

Reply:
I've simplified how we track these values in https://github.com/bytecodealliance/wasmtime/pull/6067/commits/2e5372c8ffb430e75442ef33cc5cd7a3b408800e and also added more extensive documentation. Hopefully this clarifies the code.